Create the Chado relationship field

dsenalik opened the pull request:
# New Feature

### Closes #1414 
### Depends on #2190

## Description
This creates the chado relationship field, the Tripal 4 implementation of the Tripal 3 sbo_relationship field. This field uses the various relationship tables in chado to link a record to another record of the same base table, using a controlled vocabulary term to describe the relationship.

This field is a hard one, because both `subject_id` and `object_id` need to be `store_link` because one of them points to the "hosting" record, the other to the "related" record. But it can be either one. Deep in chado storage now, any `store_link` properties are overwritten with the "hosting" record, even if a value is already there. This happens in two different places. This PR will include changes to chado storage and chado records to support more than one `store_link` property, by not overwriting the value if it is already set. (Commit https://github.com/tripal/tripal/pull/2128/commits/25c7a734973a027d44ee8e04461c393a30110846)

There is another difficulty with publish, again because we have two `store_link` properties. In ChadoRecords `selectItems()` these both get added as a condition, and for both the value of the condition is set to the "hosting" record id, and so nothing is found and as a result this field cannot be published. What needed to happen is these two are combined inside an `OR` condition group.

I did not add a field to organism in the yaml config, because the relationship field uses only one table column, and which one will depend on how a site uses organism. Ideally the `abbreviation` field would always be populated and that field could be used, but because it might not be, it is safer to make a site admin add the field.

A point for discussion, for discovery on publication, the column used is `title` if the default token string is present, however in the yaml, I used the `uniquename` column. I did so because that has a not null constraint, where the title can be left blank. It will be unlikely to have publications without titles, so let me know if the yaml should also use `title`.

 - [x] Implement value or rank when present = commit https://github.com/tripal/tripal/pull/2128/commits/ec27a5087d64624c4fb7549407cf237992e92b05
 - [x] Reverse toggle works
 - [x] CV Term synonyms are supported for the autocomplete = commit https://github.com/tripal/tripal/pull/2128/commits/a87ded9322929f196dfdbd267f05cb22cc3727e2
 - [x] Add field to yaml as appropriate for all content types
 - [x] Probably an update hook to add the field would be helpful
 
### Example for possible use in documentation
When a record is related to another record of the same type, you may use the "Chado Relationship" field to describe how the records are related. For example, you may have several publications that are chapters in an encompassing book, or books that are part of a series. Here we are specifying three publications that are part of a larger encompassing work
![relationship_1](https://github.com/user-attachments/assets/9698043a-4b00-45ef-8a2c-a28cce1a93af)

And once entered, they are displayed in this manner
![relationship_2](https://github.com/user-attachments/assets/9159bacc-d634-4a19-a743-de480ab63046)

On the related record, the field looks like this after republishing
![relationship_3](https://github.com/user-attachments/assets/1200487b-23e1-4fc7-992b-af5938938bb1)



## Testing?

Build on 4.x and switch to this branch to test the update hook
```
# drush updatedb
 -------------- ----------- --------------- ----------------------------- 
  Module         Update ID   Type            Description                  
 -------------- ----------- --------------- ----------------------------- 
  tripal_chado   10414       hook_update_n   10414 - Adds the chado       
                                             relationship field, PR 2128  
 -------------- ----------- --------------- ----------------------------- 


 ┌ Do you wish to run the specified pending updates? ───────────┐
 │ Yes                                                          │
 └──────────────────────────────────────────────────────────────┘

>  [notice] Update started: tripal_chado_update_10414
>  [notice] Added field, "analysis_relationship", to content type: "analysis".
>  [notice] Added field, "project_relationship", to content type: "project".
>  [notice] Added field, "contact_relationship", to content type: "contact".
>  [notice] Added field, "pub_relationship", to content type: "pub".
>  [notice] Update completed: tripal_chado_update_10414
 [success] Finished performing updates.
```

### Publish test
1. Create an organism
2. Create an analysis
3. Import the "Genomic content types" collection
6. Import this example file
[issue1824.gff3.txt](https://github.com/tripal/tripal/files/15109905/issue1824.gff3.txt)
For default landmark type, I used `chromosome (SO:0000340)`
7. Run the job `drush trp-run-jobs --username=drupaladmin --root=/var/www/drupal/web`
These four relationships will have been created: mRNA `part_of` gene; exon `part_of` mRNA, CDS `part_of` mRNA; polypeptide `derives_from` mRNA
9. Publish both Gene and mRNA content types and run the jobs
10. Whichever you did first, you will need to republish so that entity links appear in the relationships.

### Entry through UI test
1. Create two or more records of project, or any other content type with a relationship table (_i.e._ any core content type except for assay, array design, featuremap, phylotree, protocol, study)
2. Either while adding, or when editing, add one or more relationships, using various terms and with "Reverse" checked sometimes.
3. Check the chado table `SELECT * FROM chado.project_relationship;` to verify records are added in the correct orientation, or removed for later steps.
4. Check the drupal field table `SELECT * FROM tripal_entity__project_relationship;` to verify records are added or removed.
5. Test changing the term of an existing relationship and saving, check display and check tables.
6. Test changing the related record of an existing relationship and saving
7. Test changing the "Reverse" status of an existing relationship and saving
8. Test removing the values from an existing relationship and saving. Removing the term but not the related record should cause a validation error. Removing the related record but not the term, or removing both, should delete the record.
11. Test removing a record with the "Remove" button and saving.
12. As of commit https://github.com/tripal/tripal/pull/2128/commits/a87ded9322929f196dfdbd267f05cb22cc3727e2 using a term synonym in the autocomplete is supported, try using `IS (SO:0000973)`. When saved the base term `insertion_sequence (SO:0000973)` is used.

dsenalik committed:
early draft nonfunctional

dsenalik committed:
Merge branch '4.x' into tv4g1-issue1414-relationship-field

dsenalik committed:
Merge remote-tracking branch 'origin/4.x' into tv4g1-issue1414-relationship-field

dsenalik committed:
Merge remote-tracking branch 'origin/4.x' into tv4g1-issue1414-relationship-field

dsenalik committed:
fix undefined  in type field

dsenalik committed:
intermediate state, lots of debugging

dsenalik committed:
fix undefined $bundle in type field part 2

dsenalik committed:
not working yet

dsenalik committed:
store_link will now not overwrite existing value

dsenalik committed:
closer to working

dsenalik committed:
Merge remote-tracking branch 'origin/4.x' into tv4g1-issue1414-relationship-field

dsenalik commented:
### Relationship table documentation

Although the relationship tables have a standard naming scheme, `base_table` . `_relationship`, the columns are not so standardized. Here is the state currently. The additional columns are all nullable or have a default value except for the one case noted.

|Relationship Table|fkey naming method|other columns|
|---|---|---|
|acquisition_relationship|subject_id|value, rank|
|analysis_relationship|subject_id|value, rank|
|biomaterial_relationship|subject_id|no|
|cell_line_relationship|subject_id|no|
|contact_relationship|subject_id|no|
|cvterm_relationship|subject_id|no|
|element_relationship|subject_id|value, rank|
|elementresult_relationship|subject_id|value, rank|
|feature_relationship|subject_id|value, rank|
|library_relationship|subject_id|no|
|nd_reagent_relationship|subject_reagent_id|no|
|organism_relationship|subject_id|rank|
|phylonode_relationship|subject_id|value, rank, phylotree_id (not null)|
|project_relationship|subject_project_id|no|
|pub_relationship|subject_id|no|
|quantification_relationship|subject_id|no|
|stock_relationship|subject_id|value, rank|

dsenalik committed:
add verb; support non-standard relationship columns

dsenalik committed:
move widget fields to a single row

dsenalik committed:
new function getBundleLabel() and tests

dsenalik committed:
Now use bundle label in formatter

dsenalik committed:
optomizations

dsenalik committed:
working on remove button

dsenalik committed:
move some code to markForDeletion()

laceysanderson committed:
Merge branch '4.x' into tv4g1-issue1414-relationship-field

dsenalik committed:
change list() to []

dsenalik committed:
only type_id is a delete_if_empty trigger

dsenalik committed:
fix incorrect property type

dsenalik committed:
implement orConditionGroup for multiple store_link

dsenalik committed:
remove debugging code from widget

dsenalik committed:
only use orConditionGroup when link count > 1

dsenalik committed:
implement value and rank fields

dsenalik committed:
remove temp file

dsenalik committed:
minor edits

dsenalik committed:
Merge remote-tracking branch 'origin/4.x' into tv4g1-issue1414-relationship-field

dsenalik committed:
Merge remote-tracking branch 'origin/4.x' into tv4g1-issue1414-relationship-field

dsenalik committed:
Merge remote-tracking branch 'origin/tv4g1-issue2146-streamlineUrlAliasTitleSave' into tv4g1-issue1414-relationship-field

dsenalik committed:
Merge remote-tracking branch 'origin/4.x' into tv4g1-issue1414-relationship-field

dsenalik committed:
fix OR condition group logic

dsenalik committed:
minor formatting

dsenalik committed:
Merge remote-tracking branch 'origin/4.x' into tv4g1-issue1414-relationship-field

dsenalik committed:
remove unused link_count

dsenalik committed:
Merge branch 'tv4g1-issue2157-widget-remove-button' into tv4g1-issue1414-relationship-field

dsenalik committed:
handle removal by clearing the related record in the form

dsenalik committed:
storeLink is or_conditions aware

dsenalik committed:
simpler way to store or conditions

dsenalik committed:
remove for a more compact form

dsenalik committed:
reduce complexity

dsenalik committed:
ChadoCVTermAutocompleteController getCVtermId() now supports synonyms

dsenalik committed:
no wsod if initial values missing

dsenalik committed:
discovery will use title format to find column, e.g. pub uses "title"

dsenalik committed:
support double parentheses

dsenalik committed:
chose not choose

dsenalik committed:
revert unintended white space change

dsenalik committed:
remove n_store_link, it wasn't needed

dsenalik committed:
add type to function parameter

dsenalik committed:
draft update hook

dsenalik committed:
add all relationship fields to yaml

dsenalik committed:
finish update hook

dsenalik committed:
fix wrong storage setting default

dsenalik committed:
fix merge conflict

dsenalik committed:
Merge remote-tracking branch 'origin/4.x' into tv4g1-issue1414-relationship-field

laceysanderson commented on tripal/src/Services/TripalEntityLookup.php:
```diff
@@ -223,6 +230,34 @@ protected function getEntityIdFromRecordId($record_id, $bundle_id, $entity_type)
   }
 
+  /**
+   * Given an entity ID value, return the human-readable bundle label.
```
```suggestion
   * Retrieve the human-readable type (i.e. TripalEntityType label) for a given TripalEntity object.
```

laceysanderson commented on tripal/src/Services/TripalEntityLookup.php:
```diff
@@ -223,6 +230,34 @@ protected function getEntityIdFromRecordId($record_id, $bundle_id, $entity_type)
+   * Given an entity ID value, return the human-readable bundle label.
+   *
+   * @param int $entity_id
+   *   The entity ID value. This might be -1 for an unpublished entity.
```
```suggestion
   *   The TripalEntity ID value. This might be -1 for an unpublished entity.
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.expression_chado.yml:
```diff
@@ -113,6 +113,35 @@ fields:
+        content_type: biosample
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other analyses with relationships to this record"
```
```suggestion
        description: "Samples related to this sample. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current sample is the subject; chose Reverse if the object is the current sample."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.general_chado.yml:
```diff
@@ -487,6 +487,35 @@ fields:
+        content_type: analysis
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other analyses with relationships to this record"
```
```suggestion
        description: "Analyses related to this one. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current analysis is the subject; chose Reverse if the object is the current analysis."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.general_chado.yml:
```diff
@@ -709,6 +738,35 @@ fields:
+        content_type: project
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other projects with relationships to this record"
```
```suggestion
        description: "Projects related to this one. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current project is the subject; chose Reverse if the object is the current project."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.general_chado.yml:
```diff
@@ -964,6 +1022,35 @@ fields:
+        content_type: contact
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other contacts with relationships to this record"
```
```suggestion
        description: "Contacts related to this one. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current contact is the subject; chose Reverse if the object is the current contact."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.general_chado.yml:
```diff
@@ -1969,6 +2056,35 @@ fields:
+        content_type: pub
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other publications with relationships to this record"
```
```suggestion
        description: "Publications related to this one. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current publication is the subject; chose Reverse if the object is the current publication."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.genetic_chado.yml:
```diff
@@ -384,6 +384,35 @@ fields:
+        content_type: QTL
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other features with relationships to this record"
```
```suggestion
        description: "Sequence features related to this QTL. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current QTL is the subject; chose Reverse if the object is the current QTL."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.genetic_chado.yml:
```diff
@@ -759,6 +788,35 @@ fields:
+        content_type: sequence_variant
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other features with relationships to this record"
```
```suggestion
        description: "Sequence features related to this variant. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current variant is the subject; chose Reverse if the object is the current variant."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.genetic_chado.yml:
```diff
@@ -1134,6 +1192,35 @@ fields:
+        content_type: genetic_marker
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other features with relationships to this record"
```
```suggestion
        description: "Sequence features related to this genetic marker. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current genetic marker is the subject; chose Reverse if the object is the current genetic marker."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.genetic_chado.yml:
```diff
@@ -1509,6 +1596,35 @@ fields:
+        content_type: phenotypic_marker
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other features with relationships to this record"
```
```suggestion
        description: "Sequence features related to this heritable phenotypic marker. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current heritable phenotypic marker is the subject; chose Reverse if the object is the current heritable phenotypic marker."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.genomic_chado.yml:
```diff
@@ -192,6 +192,35 @@ fields:
+        content_type: gene
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other features with relationships to this record"
```
```suggestion
        description: "Sequence features related to this gene. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current gene is the subject; chose Reverse if the object is the current gene."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.genomic_chado.yml:
```diff
@@ -624,6 +653,35 @@ fields:
+        content_type: mrna
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other features with relationships to this record"
```
```suggestion
        description: "Sequence features related to this mRNA. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current mRNA is the subject; chose Reverse if the object is the current mRNA."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.genomic_chado.yml:
```diff
@@ -1630,6 +1688,35 @@ fields:
+        content_type: genome_assembly
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other analyses with relationships to this record"
```
```suggestion
        description: "Analyses related to this genome assembly. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current genome assembly is the subject; chose Reverse if the object is the current genome assembly."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.genomic_chado.yml:
```diff
@@ -1930,6 +2017,35 @@ fields:
+        content_type: genome_annotation
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other analyses with relationships to this record"
```
```suggestion
        description: "Analyses related to this genome annotation. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current genome annotation is the subject; chose Reverse if the object is the current genome annotation."
```
I edited this one, you had Sequence features 😁

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.genomic_chado.yml:
```diff
@@ -2094,6 +2210,35 @@ fields:
+        content_type: genome_project
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other projects with relationships to this record"
```
```suggestion
        description: "Projects related to this genome project. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current genome project is the subject; chose Reverse if the object is the current genome project."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.germplasm_chado.yml:
```diff
@@ -167,6 +167,35 @@ fields:
+        content_type: germplasm
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other stocks with relationships to this record"
```
```suggestion
        description: "Germplasm related to this one. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current germplasm is the subject; chose Reverse if the object is the current germplasm."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.germplasm_chado.yml:
```diff
@@ -410,6 +439,35 @@ fields:
+        content_type: breeding_cross
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other stocks with relationships to this record"
```
```suggestion
        description: "Germplasm related to this one. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current germplasm is the subject; chose Reverse if the object is the current germplasm."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.germplasm_chado.yml:
```diff
@@ -653,6 +711,35 @@ fields:
+        content_type: germplasm_variety
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other stocks with relationships to this record"
```
```suggestion
        description: "Germplasm related to this one. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current germplasm is the subject; chose Reverse if the object is the current germplasm."
```

laceysanderson commented on tripal_chado/config/install/tripal.tripalfield_collection.germplasm_chado.yml:
```diff
@@ -896,6 +983,35 @@ fields:
+        content_type: ril
+        label: Relationship
+        type: chado_relationship_type_default
+        description: "Other stocks with relationships to this record"
```
```suggestion
        description: "Germplasm related to this RIL. Think of relationship statements as sentences: SUBJECT TYPE OBJECT (e.g. 'Sara (subject) is the mother of (type) Jacob (object)'). By default the current RIL is the subject; chose Reverse if the object is the current RIL."
```

laceysanderson reviewed (commented):
Some suggestions to the code so far. Mostly just more help text for the relationship fields as they often take some training to be able to use.

dsenalik committed:
Apply suggestions from code review

I edited a mistake in one of your suggestions, but great improvements, thanks.

Co-authored-by: Lacey-Anne Sanderson <lacey.sanderson@usask.ca>

laceysanderson committed:
Merge branch '4.x' into tv4g1-issue1414-relationship-field

laceysanderson reviewed (approved):
Thank you for all your work on this @dsenalik! 😍 

✅ Finished the code review and everything looks good 👍 I approve of the changes to chadostorage and chadorecords.
✅ Tested the update hook and it added the fields as expected
✅ Tested publishing relationships from a recently ported version of the KnowPulse chado and they all looked great! Checked stock, feature and project based records.
✅ Tested editing a previously published relationship. The changes held and were saved to the database correctly.
✅ Tested adding and editing relationships of various content types through the UI and it worked as expected.

I think this is ready to merge once we get the brittle tests passing! 🎉
